source/lib.ts

```diff
@@ -0,0 +1,312 @@
+import MemoryStore from './memory-store.js'
+
+/**
+ * Type guard to check if a store is legacy store.
+ *
+ * @param store {LegacyStore | Store} - The store to check
+ *
+ * @return {boolean} - Whether the store is a legacy store
+ */
+const isLegacyStore = (store: LegacyStore | Store): store is LegacyStore =>
+	// Check that `incr` exists but `increment` does not - store authors might want
+	// to keep both around for backwards compatibility.
+	typeof (store as any).incr === 'function' &&
+	typeof (store as any).increment !== 'function'
+
+/**
+ * Converts a legacy store to the promisified version.
+ *
+ * @param store {LegacyStore | Store} - The legacy store or even a modern store
+ *
+ * @returns {Store} - The promisified version of the store
+ */
+const promisifyStore = (passedStore: LegacyStore | Store): Store => {
+	if (!isLegacyStore(passedStore)) {
+		// It's not an old store, return as is
+		return passedStore
+	}
+
+	// Why can't Typescript understand this?
+	const legacyStore = passedStore
+
+	class PromisifiedStore implements Store {
+		async increment(key: string): Promise<IncrementResponse> {
+			return new Promise((resolve, reject) => {
+				legacyStore.incr(
+					key,
+					(
+						error: Error | undefined,
+						totalHits: number,
+						resetTime: Date | undefined,
+					) => {
+						if (error) reject(error)
+						resolve({ totalHits, resetTime })
+					},
+				)
+			})
+		}
+
+		async decrement(key: string): Promise<void> {
+			return Promise.resolve(legacyStore.decrement(key))
+		}
+
+		async resetKey(key: string): Promise<void> {
+			return Promise.resolve(legacyStore.resetKey(key))
+		}
+
+		async resetAll(): Promise<void> {
+			if (typeof legacyStore.resetAll === 'function')
+				return Promise.resolve(legacyStore.resetAll())
+		}
+	}
+
+	return new PromisifiedStore()
+}
+
+/**
+ * Adds the defaults for options the user has not specified.
+ *
+ * @param options {Options} - The options the user specifies
+ *
+ * @returns {Options} - A complete configuration object
+ */
+const parseOptions = (passedOptions: Partial<Options>): Options => {
+	// Now add the defaults for the other options
+	const options: Options = {
+		windowMs: 60 * 1000,
+		store: new MemoryStore(),
+		max: 5,
+		message: 'Too many requests, please try again later.',
+		statusCode: 429,
+		legacyHeaders: passedOptions.headers ?? true,
+		standardHeaders: passedOptions.draft_polli_ratelimit_headers ?? false,
+		requestPropertyName: 'rateLimit',
+		skipFailedRequests: false,
+		skipSuccessfulRequests: false,
+		requestWasSuccessful: (_request, response) => response.statusCode < 400,
+		keyGenerator: (request) => {
+			if (!request.ip) {
+				console.error(
+					'WARN | `express-rate-limit` | `request.ip` is undefined. You can avoid this by providing a custom `keyGenerator` function, but it may be indicative of a larger issue.',
+				)
+			}
+
+			return request.ip
+		},
+		skip: () => false,
+		handler: (_request, response, _optionsUsed: Options) =>
```

Comment:
```suggestion
		handler: (_request, response, _next, options) =>
```

The third argument is a `next()` function, although it isn't used here. The 4th is the options. (It was a late addition, so it was tacked onto the end for back-compat.)

Also, might as well use the passed-in options now that we have them. Make it easier for someone else to copy-paste this method as the basis for their custom version.